Fix escape-keys by preserving input if invalid

vdemeester opened the pull request:
Currently, using a custom detach key with an invalid sequence, eats a part of the sequence, making it weird and difficult to enter some key sequence 🐯.

This fixes by keeping the input read when trying to see if it's the key sequence or not, and "writing" then is the key sequence is not the right one, preserving the initial input 🐹.

```
$ docker run --detach-keys=a,b,c -it busybox /bin/sh
/ #
# press a and a on the keyboard
/ # aa
/ #
# press a, b and d
/ # abd
# press a, b and c
# escape \o/
```

Fixes #21769

/cc @thaJeztah @cpuguy83  @LK4D4 @runcom @calavera 

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

duglin commented:
+1 see: https://github.com/docker/docker/pull/15666#issuecomment-165482070

thaJeztah commented:
Heh, it works. Bit of a weird side-effect is that typing `a` twice, will not output the first `a`, until the second `a` is pressed. Guess that's ok, because using `a,b,c` as detach-key is not really a smart choice

duglin commented:
any chance of getting some testcases?

vdemeester commented:
@duglin I'm thinking of it, but I'm not sure how still... :sweat:

thaJeztah commented:
Perhaps you can add the "test" case to the commit message at least

duglin commented:
@vdemeester you may get some inspiration from the tests that use a tty - e.g. TestRunAttachDetach

vdemeester commented:
@duglin oh you are right. Remembered the way we pty to send input and detach the container but I did not remembered if we check the output content or not… and it seems we do `\o/`. Test incoming :angel:.

vdemeester committed:
Fix escape-keys by preserving input if invalid

Currently, using a custom detach key with an invalid sequence, eats a
part of the sequence, making it weird and difficult to enter some key
sequence.

This fixes by keeping the input read when trying to see if it's the key
sequence or not, and "writing" then is the key sequence is not the right
one, preserving the initial input.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
Updated with tests :wink:

LK4D4 commented:
LGTM
ping @duglin

calavera commented:
LGTM

duglin commented on container/container.go:
```diff
@@ -484,7 +484,9 @@ func copyEscapable(dst io.Writer, src io.ReadCloser, keys []byte) (written int64
 		nr, er := src.Read(buf)
 		if nr > 0 {
 			// ---- Docker addition
+			preservBuf := []byte{}
```
Trying to follow the flow here, but it seems to me that "preservBuf" shouldn't be needed since it MUST match the keys, right?  meaning if we're on the 3rd char of 'keys' then preservBuf must be the same as the first 2 chars of keys, right?

vdemeester replied:
@duglin hum it matches a part of the key, true, until the last *wrong* key is entered. Could have done something with `keys`, `i` and the latest read but it failed a little less readable to me :sweat_smile:

duglin replied:
yea, I just wanted to avoid all of the alloc's and appends as we work our way thru the input.
